### DISCUSSION TITLE
Understanding live exports rules

### DISCUSSION THREAD

**Discussion by tmpfs** [endojs/endo#1951] [General]
Just a general discussion to help me understand the rules for live exports. So far I think that the following are live exports:

* Late assignment to an exported variable
* Re-export from an imported module
* Exported function expression
* Exported class declaration

Are there other cases where I should transform as a live export?

Regarding the exported function expression it appears that the `Object.defineProperty()` and `$h‍_live.fn($c‍_fn);` are hoisted to the top; currently my transform is not doing this ([input](https://github.com/LavaMoat/basalt/blob/main/tests/transform/export-name-6/input.js) & [output](https://github.com/LavaMoat/basalt/blob/main/tests/transform/export-name-6/output.js)) - I assume it is required?. See the output of [test case 6](https://github.com/endojs/endo/blob/master/packages/static-module-record/TESTS.md#case-6) for the output with hoisting.

Also I am a bit confused by [test case 7](https://github.com/endojs/endo/blob/master/packages/static-module-record/TESTS.md#case-7). Why is it calling `$h‍_live.C(C);` when the exported variable is `count` I would expect it to operate on the exported variable not the class definition. Any pointers much appreciated!

**Comment by tmpfs** [comment]
Apologies, I mis-read the output program for [test case 7](https://github.com/endojs/endo/blob/master/packages/static-module-record/TESTS.md#case-7) I see now that both the variable `count` and the class `C` are treated as live exports.

I didn't realize that exported classes can be re-assigned, this was unexpected for me:

```javascript
export class C {}

C = function foo() {
  console.log('baz');
}

C()
```

**Comment by erights** [comment]
Test case 7 is
```js
export let count = 0;
export class C {} if (C) { count += 1; }
```
I see why `count` is a live binding --- it is assigned to.
But why is `C` considered a live binding? It is not assigned to.

**Comment by tmpfs** [comment]
@erights, it seems that exported classes can be re-assigned (which I didn't know) however you are right that in Test Case 7 there is no assignment and it would be safe to skip the live export IMO.

This would be a case where we would need a live export I think (`C.mjs`):

```javascript
export class C {};

C = function() {
  console.log('baz')
}
```

Then import it and the function assignment which overrides the class is used:

```javascript
import {C} from './C.mjs'
C()
```
